src/app/InteractionModelEngine.cpp

```diff
@@ -190,7 +190,6 @@ CHIP_ERROR InteractionModelEngine::OnUnknownMsgType(Messaging::ExchangeContext *
     // err = SendStatusReport(ec, kChipProfile_Common, kStatus_UnsupportedMessage);
     // SuccessOrExit(err);
 
-    apExchangeContext->Close();
```

Comment:
Since an unknown message is received here, how we know the exchange is auto closed in this case since we don't know if kFlagWillSendMessage flag is set for an unknown message

Reply:
I'm not sure I follow.  We're in an unsolicited message handler.  The only way we land in this code is if an unsolicited message came in off the network.  This is therefore a brand-new exchange that was just created.  We're not telling anyone anything about this exchange; just silently closing it, before this PR.  After this PR we continue silently closing it, since none of the conditions that would cause it to not close apply in this case.

Reply:
OnMessageReceived is used as both unsolicited message handler and received message handler. why the only way to land in this code is an unsolicited message came in off the network?

1. Suppose InvokeCommandRequest is first received, and an new exchange is created by ExchangeMgr and handle over InvokeCommandRequest to OnMessageReceived, then we should fall under OnInvokeCommandRequest, right?

2. Suppose the other end send an message other than InvokeCommandRequest/ReadRequest/WriteRequest over an existing exchange, we could also fall under OnUnknownMsgType, right?

Reply:
> OnMessageReceived is used as both unsolicited message handler and received message handler.

Hmm...   I thought for some reason that IMEngine reset the delegate to the object it was handing off to, but it clearly does not....

But the reason I had thought that is that what the code _actually_ does is totally unsafe.  Let's say that we get a Read Request message.  We allocate a ReadHandler, call OnReadRequest on it, that goes async waiting to send a response on the exchange (this is with or without my changes; the difference is that with my changes we call `WillSendMessage` when we go async).

If now an unknown message comes in on the same exchange, the code before my change will close the exchange, and now the ReadHandler has a dangling pointer to a deleted exchange, and will crash when it tries to send the read response.  This is exactly the sort of thing I am trying to fix by switching from manual exchange management to telling the exchange what you plan to do with it....

Back to your specific scenario: That one is actually ok with or without my changes, for the moment, but only because `OnInvokeCommandRequest` synchronously replies and closes the exchange.  That means that (outside of the loopback setup in tests) the second message, which would go to `OnUnknownMsgType`, never gets there, because the exchange is already closed, its delegate nulled out, and hence we can't land in `OnMessageReceived`.  Once commands start doing async work it looks more like the Read Request case above.

Now the argument might be that if we got a message the spec did not intend on an exchange we should close it.  But that's also true if an exchange that started with an Invoke Request suddenly gets a Read Request.  That is not an expected message on that exchange, and it should be treated like any other random unexpected message.

Probably the simplest way to make that work is in fact to make ReadHandler/CommandHandler/WriteHandler into ExchangeDelegates and have them take over all message-handling on the exchange until the decide they are done with it and close it.  I am happy to do that as a followup to this PR, or as prereq if desired.

Reply:
Ack